### ISSUE TITLE
Latest 4.0.0-beta.1 hangs

### ISSUE THREAD

**Issue by creage** [amireh/happypack#166]
After upgrade to the latest beta, some of our builds started runing for hours, keeping node.js instances up, with zero activity. No errors fired. Can't identify the reason, but looks like some callbacks or exit codes are not processed correctly.

Downgrading to the 3.1.0 fixes the issue.

**Comment by amireh** [comment]
Hmm, this isn't on Windows I assume? Can you share your build config by any chance?

**Comment by creage** [comment]
It is on Windows. Build config is quite simple, and looks like these

```javascript
const HappyPack = require('happypack');

HappyPack.SERIALIZABLE_OPTIONS = HappyPack.SERIALIZABLE_OPTIONS.concat(['postcss']);

module: {
    rules: [
        {
            test: /\.js$/,
                exclude: /node_modules/,
                use: 'happypack/loader?id=js'
        },			
        {
            test: /\.po$/,
            use: 'happypack/loader?id=po'
        },
        {
            test: /\.css$/,
            use: ExtractTextPlugin.extract({
                fallback: 'style-loader',
                use: 'happypack/loader?id=css'
            })
        },
        {
            test: /\.less$/,
            exclude: /node_modules/,
            use: ExtractTextPlugin.extract({
                fallback: 'style-loader',
                use: 'happypack/loader?id=less'
            })
        }
    ]
},	
plugins: [
    ...
    new HappyPack({
        id: 'js',
        loaders: ['babel-loader'],
        verbose: false
    }),
    new HappyPack({
        id: 'po',
        loaders: ['json-loader', 'po-loader'],
        verbose: false
    }),
    new HappyPack({
        id: 'css',
        loaders: ['css-loader?importLoaders=1', 'postcss-loader'],
        verbose: false
    }),
    new HappyPack({
        id: 'less',
        loaders: ['css-loader?importLoaders=1', 'postcss-loader', 'less-loader'],
        verbose: false
    })
    ...
]
```

**Comment by amireh** [comment]
Thanks, I'll see if I can reproduce on a Windows 7 VM.

**Comment by creage** [comment]
I want to add, that 3.1.0 also started to hang randomly, so, it should be something related to a Node 8 compatibility, I believe. Because we see some of NPM modules failing due to API changes in the latest NodeJS.

**Comment by MrZhongFF** [comment]
It occurs randomly。But my  Node version is 6.9.4。

**Comment by stooboo** [comment]
for me  (windows + node v6.11.0)
3.1.0 works fine :-)
4.0.0-beta.1 hangs :-(

**Comment by abergs** [comment]
Same here.
Windows 10.
Node 7.10
Webpack 3.0

**Comment by hxfdarling** [comment]
I have also encountered

**Comment by hiyangguo** [comment]
I also appeared this problem. node version is 8.2.1.

**Comment by initialdmg** [comment]
i had this problem too.
node 8.2.1
webpack 2.7.0
windows 7 64bit
happypack 4.0.0

**Comment by carpogoryanin** [comment]
+
Sometimes this problem occurred.

Windows 10 64bit
Node.js v7.5.0
webpack 3.3.0
happypack 4.0.0-beta.1

**Comment by Saltallica** [comment]
Same. Sometimes it works, sometimes... not so much

Win 10 (64bit)
Node 6.11.1
webpack 2.6.1
happypack 4.0.0-beta.2

**Comment by hxfdarling** [comment]
Just use 3.x. It's works

**Comment by amireh** [comment]
This must be Windows exclusive because we've been using v4 under node v8.2.1 on both Linux and OS X with no issues.

Is there any way you can reproduce this with a project I can look at?

**Comment by creage** [comment]
@amireh it is so random, that is close to be not reproducible. I've tried to create a new repo, having config similar to my prod one, but with very small code base, and I've succeed in reproducing the issue only once per like 50 runs. But on my original project it happens quite often - it hangs every 5-7 run. I'm not sure if it depends on the project size..., and have no idea how to provide you a repro steps.

**Comment by amireh** [comment]
Okay, i was able to reproduce this. The culprit seems in that we're no longer storing the result to disk in the background process then reading it from disk in the foreground (which is the behaviour in v3). I suspect this is related to the IPC channel buffers but I couldn't figure out a solution yet.

But this is progress, at least we know what's going on now. In the worst-case scenario, we can make using the filesystem an option to be toggled on for Windows (or make that implicit) but I'm hoping we can find a way to make this work cross-platform without resorting to fs.

**Comment by amireh** [comment]
This should be fixed in the `4.0.0-beta.4` release. :balloon: Feel free to re-open if this is still an issue (or please don't :joy:).
